Content/posts/otel-integration.md

```diff
@@ -0,0 +1,303 @@
+2. The second one is the queue workers periodically pushing the data to a collector, and the Prometheus instance scraping the collector.
+
+#### HTTP Server
+
+This first approach is the simplest one, because all we're doing is opening up a `/metrics` endpoint which Prometheus is configured to scrape on its own, every x seconds. This also allows Vapor to ignore the Prometheus instance completely and really just treat it as a normal client, which is good, especially in a distributed system, as we don't want to burden the API more than we have to. However, the catch is obviously that we need to be running an HTTP server to handle the requests.
+
+To implement this approach, we use the [Swift Prometheus](https://github.com/swift-server/swift-prometheus) package. This package is really simple: all it does is collect and export the SwiftMetrics emitted metrics in a Prometheus format. First things first we'll want to add the package to our application in the `Package.swift` file, both in the package's dependencies and in the target's dependencies:
+
+```swift
+.package(url: "https://github.com/swift-server/swift-prometheus.git", from: "2.0.0")
+```
+```swift
+.product(name: "Prometheus", package: "swift-prometheus")
+```
+
+Once that's done, we can configure the collector:
+
+```swift
+let factory = PrometheusMetricsFactory()
+MetricsSystem.bootstrap(factory)
+```
+
+And finally add the metrics-emitting route:
+
+```swift
+router.get("metrics") { _ in
+	var buffer = [UInt8]()
+	(MetricsSystem.factory as? PrometheusMetricsFactory)?.registry.emit(into: &buffer)
+	return String(bytes: buffer, encoding: .utf8) ?? ""
+}
+```
+
+That's it! Once we build and run we should see some metrics being printed out. This endpoint will be used by our Prometheus instance.
+
+To configure Prometheus, we need to create a `prometheus.yml` configuration file in the same directory as the `docker-compose.yml` file. This is the initial configuration:
+
+```yaml
+scrape_configs:
+  - job_name: 'api'
+    scrape_interval: 30s
+    static_configs:
+      - targets: ["host.docker.internal:8080"]
+```
+
+This configuration tells Prometheus to scrape the `/metrics` endpoint on the host machine every 30 seconds.
+
+> Note: The `host.docker.internal` address is a special address that Docker uses to refer to the host machine from within a container. This supposes that the Vapor application is running on the host machine and not in a container.
+
+Then we can add Prometheus to our `docker-compose.yml` file:
+
+```yaml
+services:
+  prometheus:
+    image: prom/prometheus
+    volumes:
+      - ./prometheus.yml:/etc/prometheus/prometheus.yml
+    ports:
+      - 9090:9090
+```
+
+This should be enough to get Prometheus up and running and scraping the Vapor application.
+We'll update both the `docker-compose.yml` and the `prometheus.yml` files later on to include the queue workers.
```

Comment:
Let's add a note here to talk about sidecars that most people will use in a production app